enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,642 @@
+
+
+#### Binary Discovery
+The existence of test extension binaries can be registered one of two ways by test authors. 
+
+##### OpenShift Payload Extension Binaries
+For OpenShift payload components contributors can advertise the existence of an extension binary
+by adding information (the imagestream tag for the OCP payload component and the path to the binary 
+within their image) to a simple registry datastructure in github.com/openshift/origin. 
+
+##### Non-Payload Extension Binaries
+For non-payload components, contributors must advertise their extension binary for `openshift-tests`
+to discover. This is accomplished by creating an ImageStream / ImageStreamTag with a special label:
+
+`testextension.redhat.io/component=<product-name>-<component-name>`
+
+An annotation on the ImageStreamTag will then identify the binary to extract and (optional)
+arguments to pass ahead of the extension interface verbs.
+`testextension.redhat.io/binary=<binary-path>.gz [--argument=value]`
+
+The creation of this imagestream/label/annotation should be gated on the existence of the
+`TestExtensionAdmission` custom resource definition (the code should not attempt to 
+inspect `TestExtensionAdmissions` -- just check for the CRD itself).
+
+Optional Operator authors must ensure that the image carrying the extension binary
+is identified in their ClusterServiceVersion (CSV) so that tools like `oc-mirror`
+will copy image(s) bearing extension binaries to disconnected clusters.
+
+Instances of `TestExtensionAdmission` gate which extension binaries `openshift-tests`
+will use for test discovery. This is because extension binaries will be invoked
+with a `system:admin` kubeconfig when tests are run. Administrators
+must opt-in to this risk by instantiating one or more `TestExtensionAdmission`
+objects. 
+
+`TestExtensionAdmission` contain a list of namespaces/imagestreams 
+from which `openshift-tests` is permitted to extract extension binaries. This
+can include a wildcard patterns for namespace, imagestream, or both.
+
+```yaml
+kind: TestExtensionAdmission
+metadata:
+  name: example
+spec:
+  permit:
+  # All imagestreams in the openshift namespace with the testextension label
+  - "openshift/*" 
+  # All imagestreams on the cluster with the testextension label
+  - "*/*" 
+```
+
+`openshift-tests` will SKIP a single synthetic test whenever an extension binary is detected in 
+any imagestreamtag but which is not permitted for execution by a `TestExtensionAdmission` 
+instance. This will allow users to detect tests that are available but not yet explicitly
+permitted to run.
+
+#### Binary Format
+Extension binaries are extracted from container images into a running Pod executing 
+`openshift-tests`. Contributors do not necessarily know which version of RHEL the
+`openshift-tests` binary will be running on, so, for maximum portability, test
+extension binaries should be statically compiled.

```

Comment:
Do we need to statically link openshift-tests too? It's not today AFAICT.

Reply:
It's not and that's probably because FIPS tests generally need to run with dynamic linking, like most of what they're exercising. For other binaries that need to be "agnostic" they are generally compiled for both RHEL 8 and 9.

Not certain that these binaries would need to operate with FIPS enabled.

Reply:
The static linking only serves to insulate us from the issues which arise when binaries are copied from one environment to another. There is no need for openshift-tests to be statically linked for the purposes of CI since it will be running in its native container. Maybe it would be beneficial to end users who aren't building their own or running inside the `tests` container, but if we haven't received any complaints, it is probably not an issue.

> Not certain that these binaries would need to operate with FIPS enabled.

It wouldn't hurt. Being statically linked, extension binaries can't be FIPS compliant today, but we can prevent the "FIPS or Die" feature from killing them when run in a FIPS-enabled environment.

Reply:
https://github.com/openshift-eng/openshift-tests-extension/pull/13